changed dealflow bookmarking to check update_time, force UTC

alewis-ce opened the pull request:
Changed dealflow to find deals with changes based on `update_time` instead of `stage_change_time`; also forced UTC for `stream_start` variable.

timvisher commented:
How's this PR looking? Is testing progressing, etc.? What do we think is left before we can merge?

alewis-ce commented:
Yes, testing is progressing. It looks like we've eliminated discrepancies between my local run version and the Stitch version. Now we're testing a few smaller changes to remove a lag we added previously and to reduce the size of the attribution window to minimize the number of duplicate rows read by Stitch.

timvisher commented on tap_pipedrive/stream.py:
```diff
@@ -106,6 +105,7 @@ def state_is_newer_or_equal(self, current_state):
 
     def write_record(self, row):
         if self.record_is_newer_equal_null(row):
+            #logger.info('DEAL WRITTEN: {}, {}'.format(row['item_id'], row['log_time']))
```
FWIW This might be a good place to use `logger.debug`? That's off by default but can be easily flipped on during development.

alewis-ce replied:
ah, I meant to delete that line. I was just using it to print things out on my side for spot checks.

alewis-ce commented:
New update. The stream iterates over pages of deal_ids, and then iterates over each deal_id per page. When it gets to the last page of deal_ids, the more_items_in_collection flag gets set to False because there are no more deal_ids to find. The stream then skips over that deal thinking there's no information to gather from it. I force the more_items_in_collection flag to True for the first deal on each page to avoid this.

timvisher commented:
I really appreciate you're commenting on a lot of the logic in the code rather than just in the PR. Thanks! :)

alewis-ce commented:
Found another bug. I had set the attribution window to find the deals with changes, but then it was ignored when it came time to write the rows because the timestamps are checked against the initial state. I updated the way I set the bookmark, I changed the definition of the checkpoint, and I limited the types of rows that are written to avoid confusion as to what the stream contains.
I'm pretty sure this fixes all the issues I've been seeing.

alewis-ce commented:
Final commit. On our end, all discrepancies between running tap-pipedrive locally and running it through Stitch have been resolved.

luandy64 commented on tap_pipedrive/stream.py:
```diff
@@ -158,11 +159,14 @@ def get_deal_ids(self, tap):
-            
+                                    if (self.stream_start > pendulum.parse(response.json()['data'][i]['add_time']) >= checkpoint) 
+                                    or (response.json()['data'][i]['stage_change_time'] is not None and 
+                                           self.stream_start > pendulum.parse(response.json()['data'][i]['stage_change_time']) >= checkpoint)]
```
At first glance, this is a very dense and complex looking list comprehension. Can we make this more readable and less complex?

I think a helper function would make this section less complex and if we throw in some comments, we could clarify the logic too.

alewis-ce replied:
Yeah, that list comprehension was pretty dense. I stuck it into a helper function, as requested, and I split it into two pieces with comments on each.

luandy64 commented on tap_pipedrive/streams/dealflow.py:
```diff
@@ -2,18 +2,20 @@
+    id_endpoint = 'deals/{}/flow'
+    schema = 'dealflow'
+    state_field = 'log_time'
+    key_properties = ['id', ]
```
I love that we are ripping out the tabs here. Could we also make sure every other file gets converted to spaces as well?

alewis-ce replied:
All files that I have touched should now be tab free.

luandy64 commented on tap_pipedrive/stream.py:
```diff
@@ -157,12 +159,23 @@ def get_deal_ids(self, tap):
+        # find all deals that a) had a stage change at any time (i.e., the stage_change_time is not None),
+        #                     b) had a stage change after the start time and before the stop time, and
+        #                     c) are not in add_ids
+        changed_ids = [data[i]['id'] for i in range(len(data)) if (data[i]['id'] not in added_ids) & (data[i]['stage_change_time'] is not None
```
This looks great 👍 

I'm going to push back on the formatting though. Can we change it to something like this?
```
[data[i]['id']
 for i in range(len(data))
 if (data[i]['id'] not in added_ids)
 and (data[i]['stage_change_time'] is not None
      and start <= pendulum.parse(data[i]['stage_change_time']) < stop)]
```

luandy64 replied:
I don't know if changing the other one would improve its readability. So I'll let you make the call on that one.

alewis-ce commented:
List comprehension formatting is changed. Let me know if you need anything else.

luandy64 commented:
Can you squash the commits? After that I'll happily merge this!

alewis-ce commented:
Done.

luandy64 commented:
Thanks for working on this! 👍
